sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -830,11 +845,184 @@ public void populateDisplayData(DisplayData.Builder builder) {
       inner.populateDisplayData(builder);
     }
 
+    private boolean hasStatementAndSetter() {
+      return inner.getStatement() != null && inner.getPreparedStatementSetter() != null;
+    }
+
     @Override
     public PDone expand(PCollection<T> input) {
+      // fixme: validate invalid table input
+      if (input.hasSchema() && !hasStatementAndSetter()) {
+        checkArgument(inner.getTable() != null, "table cannot be null if statement is not provided");
+        Schema schema = input.getSchema();
+        List<Schema.Field> fields = getFilteredFields(schema);
+        inner = inner.withStatement(JdbcUtil.generateStatement(inner.getTable(), fields));
+        inner = inner.withPreparedStatementSetter(new AutoGeneratedPreparedStatementSetter(fields, input.getToRowFunction()));
+      }
+
       inner.expand(input);
       return PDone.in(input.getPipeline());
     }
+
+    private List<Schema.Field> getFilteredFields(Schema schema) {
+      List<Schema.Field> fields = schema.getFields();
+      Connection connection;
+      try {
+        connection = inner.getDataSourceProviderFn().apply(null).getConnection();
+      } catch (Exception e) {
+        throw new RuntimeException("Error while determining columns from table: " + inner.getTable());
+      }
+      List<String> databaseColumnList = getTableColumnList(connection);
+      fields = fields.stream().filter((field) -> {
+        return databaseColumnList.contains(field.getName().toLowerCase());
+      }).collect(Collectors.toList());
+      return fields;
+    }
+
+    private List<String> getTableColumnList(Connection connection) {
```

Comment:
Make use of `SchemaUtils.toBeamSchema` instead.

Reply:
Done.
We are matching field names in lower case and field type names only. Do you suggest we match if the nullable status of field in schema and database match too? If so, then we can just compare the fields with `.equal()` method.